### PR TITLE
Allow the join/leave message to be overridden by mods.

### PR DESCRIPTION
**Pull request by red-001**
Requested in #4227. this allows modifying join/leave messages without modifying builtin, making it possible to do in a mod.

### CODE REVIEW

**Comment by est31** [comment]
I think @sofar didn't want APIs like this, changing the minetest variable.

**Comment by paramat** [comment]
I don't think there's enough need for this :-1: 
A simple join / leave message is fine.

**Comment by nerzhul** [comment]
I think a proper API to dispatch system messages to client is needed

**Comment by Amaz1** [comment]
> You can already do this in a mod, which won't remove the auto message, however it's good enough.

Sorry, but I really don't think that this is good enough.

If one were to make a mod which allowed players to have their own chat colour, or say gave admins a different chat colour to normal players, and one wanted to make the players'/admins' join/leave message that colour, it would look really ugly, to the extent that it is not doable in a mod, because you would have two join/leave messages - one white, one coloured. This is one of a number of situations that is impossible to do neatly at the moment. (e.g. Another one could be a mod which wants to add "(Admin)" or "(Mod)" after an admin's or moderator's name in the chat or a mod which allowed certain players to join the server unannounced.)
Another possible use might be for a fantasy subgame, where instead of saying "x joined the game", the subgame developer might want to say "x entered the world of y," which currently is impossible to do without causing a massive mess. (And I say this as a subgame developer, who would be interested in doing this...)

Sorry for this wall of text :P

**Comment by paramat** [comment]
I am neutral now, this is fairly simple as it just allows the re-definition of functions.

File: builtin/game/misc.lua [hunk 1]
```diff
@@ -86,20 +86,27 @@ end
...
 	end
 	core.chat_send_all(announcement)
+end
+
+core.register_on_joinplayer(function(player)
+	local player_name = player:get_player_name()
+	player_list[player_name] = player
+	core.send_join_message(player_name)
+end)
+core.register_on_leaveplayer(function(player, timed_out)
```
Comment (rubenwardy) [inline builtin/game/misc.lua]: space above here please

**Comment by red-001** [comment]
updated and rebased

**Comment by Amaz1** [comment]
Could a dev review this please? I would find it very useful...

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2641,6 +2641,10 @@ These functions return the leftover itemstack.
     * Replaces definition of a builtin hud element
     * `name`: `"breath"` or `"health"`
     * `hud_definition`: definition to replace builtin definition
+* `minetest.send_join_message(player_name)`
+	* Can be overridden by mods to change the join message
+* `minetest.send_leave_message(player_name, timed_out)`
+	* Can be overridden by mods to change the leave message
```
Comment (SmallJoker) [inline doc/lua_api.txt]: `     * Can be overridden`

Comment (SmallJoker) [inline doc/lua_api.txt]: Missing indents: (or see documentation above)
```
     * Can be overridden [..]
````

**Comment by sofar** [comment]
> I think @sofar didn't want APIs like this, changing the minetest variable.

Correct, I really don't like API's that override parts or members of `minetest`. Ideally, we can use `luacheck` to verify that code looks OK and that nobody is pulling shenanigans that are impossible to debug (simply because we often have no idea what mod is replacing minetest.foo_function with a custom or even broken implementation.

So, I would like to get closer to, not further away from, the ideal situation where nobody has to write members of `minetest` in the first place.

**Comment by rubenwardy** [comment]
I'm not sure what the best API for this would be without overriding a minetest member. Maybe the ability to label and then delete registered callbacks?

ie:

```lua
core.default_callbacks.on_join = minetest.register_on_joinplayer(function(player)
    minetest.chat_send_all(player:get_player_name() .. " has joined!")
end)

-- elsewhere
core.unregister_on_joinplayer(core.default_callbacks.on_join)
```

**Comment by SmallJoker** [comment]
@rubenwardy The same overriding procedure happens for `core.is_protected`. Adding callbacks only for this feature is out of the scope. The current concept looks good and should be merged this way - concept approval from my side.

**Comment by rubenwardy** [comment]
well, I'm okay with this PR however I wanted to try and allay sofar's concerns

**Comment by sfan5** [comment]
@smalljoker
is that a full :+1: for this PR, meaning can this be merged?

**Comment by sfan5** [comment]
d10cccee3108b45cf7085ffdbff9b5f2a4045ea7

**Pull request closed without merging**